spec/models/booking_spec.rb

```diff
@@ -0,0 +1,180 @@
+          expect(booking.reload.status.to_sym).not_to eq(:canceled)
+        end
+      end
+    end
+
+    describe "#confirm_booking" do
+    context "when booking is pending" do
+      let(:booking) { create(:booking) }
+      it "confirms the booking and sends confirm email" do
+          expect { booking.confirm_booking }.to change { booking.reload.status.to_sym }.to(:confirmed)
+        end
+      end
+
+      context "when booking is not pending" do
+        it "does not confirm the booking" do
+          expect { booking_canceled.confirm_booking }.to raise_error(RuntimeError)
+          expect(booking_canceled.status.to_sym).not_to eq(:confirmed)
+        end
+      end
+    end
+
+    describe "#successed_booking" do
+      let(:booking) { create(:booking) }
+      it "updates the booking status to successed and sends success email" do
+        booking.confirmed!
+        expect { booking.successed_booking }.to change { booking.status.to_sym }.to(:successed)
+      end
+    end
+
+    describe "overtime_booking" do
+      it "returns pending bookings with date_start >= now" do
+        expect(Booking.overtime_booking).to include(future_pending_booking)
+        expect(Booking.overtime_booking).not_to include(past_confirmed_booking)
+      end
+    end
+
+    describe "success_booking" do
+      it "returns confirmed bookings with end_date <= 3 days ago" do
+        expect(Booking.success_booking).to include(past_confirmed_booking)
+        expect(Booking.success_booking).not_to include(future_pending_booking)
+      end
+    end
+  end
+  describe 'Scope' do
+    let!(:tour_detail1) { create(:tour_detail, max_people: 5, price: 100) }
+    let!(:tour_detail2) { create(:tour_detail, max_people: 5, price: 100) }
+    describe '.incomes' do
+      it 'returns total amount grouped by day for successful bookings within the specified date range' do
+        create(:booking_success, tour_detail: tour_detail1, created_at: Time.zone.now, numbers_people: 1)
+        create(:booking_success, tour_detail: tour_detail2, created_at: 1.days.ago + 1.hour,  numbers_people: 1)
+
+        result = Booking.incomes(1.days.ago, Time.zone.now)
+
+        expect(result).to eq({ 1.days.ago.strftime(Settings.dd_mm_yyyy_fm) => 100, Time.zone.now.strftime(Settings.dd_mm_yyyy_fm) => 100 })
+      end
+    end
+
+    describe '.new_bookings' do
+      it 'returns count of new bookings grouped by tour detail within the specified date range' do
+        create(:booking, tour_detail: tour_detail1, numbers_people: 1, created_at: Time.zone.now)
+        create(:booking, tour_detail: tour_detail2, numbers_people: 1, created_at: 1.days.ago + 1.hour)
+        create(:booking, tour_detail: tour_detail1, numbers_people: 1, created_at: Time.zone.now)
+        result = Booking.new_bookings(1.days.ago, Time.zone.now)
+        expect(result).to eq({ tour_detail1 => 2, tour_detail2 => 1})
+        #
+      end
+    end
+
+    describe '.status_bookings' do
```

Comment:
- e tìm hiểu và phân biệt các keywords: `let, context, describe` nhé
- Phân biệt `let` vs `let!`

Reply:
let : let khai báo biến trong khối và có giá trị trong từng example ( mỗi example ("it" block) sẽ khởi tạo lại )
let! : let! tương tự như let dùng để khai báo biến nhưng sẽ chỉ khởi tạo lại lần đầu (= việc chạy 1 before ngầm)
context: dùng để nhóm các điều kiện của example.
describe: dùng để nhóm các test case theo class, module, method hoặc theo action của controller.